Initial commit for comanage troubleshooting guide (SOFTWARE-5775) (#107)
osg-htc/technology: docs/software/comanage-troubleshooting-guide.md

COManage Troubleshooting Guide
==============================

A resource for COManage administrators to find solutions to commonly encountered problems.

COManage Troubleshooting Items
----
This section contains some of the issues you may encounter when interacting with the OSG COManage.

Problem list:

>  - [Valid Tiger User is unauthorized with Dex credentials](#valid-tiger-user-is-unauthorized-with-dex-credentials)

Reviewer: This is unnecessary since we use MkDocs here and it will generate a table of contents for us based on the headings

### Valid Tiger User is unauthorized with Dex credentials

Under certain circumstances Dex may not be able to match a user to their COManage identity,
 possibly due to incorrect and/or missing organization identities/identifiers.

#### Symptoms

 - Tiger-Dex credentials missing some or all of a users group permissions.
 - User unable to connect to the Tiger cluster.
 - Error message from Tiger involving an unauthorized user.
    - Example error message: 

    >kubectl get pods -n osg-dev

    >Error from server (Forbidden): pods is forbidden: User "user@email.edu" cannot list resource "pods" in API group "" in the namespace "osg-dev""
 

>  - Empty Dex response

Reviewer: ```suggestion
 - Empty groups in the Dex response after logging in with SSO
```

#### Next actions

1. Ensure that the user's LDAP record contains the attribute `uid`,
 which shares the same value as the COManage identifier used as a source for the LDAP provisioner target.
    - LDAP attributes and their COManage sources can be found [here](https://registry.cilogon.org/registry/ldap_provisioner/co_ldap_provisioner_targets/edit/6)
1. Have the user log in
1. Have the user fetch a new Dex token  

> ####Explanation of SAML -> CILogon OIDC sub claim -> LDAP matching used in Dex.

Reviewer: ```suggestion
#### Explanation of SAML -> CILogon OIDC sub claim -> LDAP matching used in Dex.
```

COmanage Troubleshooting Guide
====

> Introduction
> ----

Reviewer: ```suggestion
```

See https://osg-htc.org/technology/documentation/writing-documentation/#introductions

> The purpose of this guide is to be a resource for COmanage administrators to find solutions to commonly encountered problems.

Reviewer: ```suggestion
The purpose of this guide is to be a resource for COManage administrators to find solutions to commonly encountered problems.
```

COManage Troubleshooting Items
----
This section contains some of the issues you may encounter when interacting with the OSG COmanage.

Problem list:

 - [Tiger-Dex Unable to match User to COmanage CO Person identity](#tiger-dex-unable-to-match-user-to-comanage-co-person-identity)

> ###Tiger-Dex Unable to match User to COmanage CO Person identity

Reviewer: ```suggestion
### Tiger-Dex Unable to match User to COmanage CO Person identity
```

Whitespace for readability. Also I'm not sure if mkdocs parses the heading without it

Also, this is more of the root cause of the problem rather than the problem itself. The problem is along the lines of "User is not authorized to access Tiger with Dex credentials"

> Under certain circumstances Dex may not be able to match a user to their COmanage identity,

Reviewer: ```suggestion
Under certain circumstances Dex may not be able to match a user to their COManage identity,
```

 possibly due to incorrect and/or missing organization identities/identifiers.

> ####Symptoms

Reviewer: ```suggestion
#### Symptoms
```

> Symptoms include:
> 

Reviewer: ```suggestion
```

Superfluous since the heading is "Symptoms"

>  - Tiger-Dex credentials not including some or all of a users group permissions.

Reviewer: ```suggestion
 - Tiger-Dex credentials missing some or all of a users group permissions.
```

>  - Not being able to connect to the Tiger cluster.

Reviewer: ```suggestion
 - Being unable to connect to the Tiger cluster.
```

It'd be worth adding an example error message here. Inability to connect will manifest differently than failure to authenticate/authorize (this problem is the latter, anyway)

Author: Do you happen to still have an error message you and/or Huijin encountered when you were dealing with this issue that we could add to the doc or is a general "error message from Tiger referring to a failure to authenticate/authorize" good enough?

####Next actions

> - Ensure all of the user's org identities / identifiers required by LDAP exist and match across COmanage and LDAP.

Reviewer: Which org identities / identifiers?

    - Found at `https://registry.cilogon.org/registry/ldap_provisioner/co_ldap_provisioner_targets/edit/6`

> - Setting the missing/mismatched value manually did not seem to work, 

Reviewer: This should be an enumerated list and this second point should be that the user should log in.

From https://osg-htc.org/technology/documentation/writing-documentation/#tips-for-writing-procedural-instructions

> Number all steps (as opposed to using bullets)

>
> It retrieves configured LDAP attributes from the record it has found,
> and then sends those values out as claims with names as configured.
>
>> What determines that the uid attribute is essential for this process?
>
> Technical debt.
>
> "Real soon now" [Scott Koranda] will be upgrading COmanage Registry to version 4.1.2,
> and with that upgrade there will be a new version of the Registry plugin that allows you to configure OIDC clients.
>
> With that new version of the plugin you will be able to select an LDAP attribute other than uid to create,
> for example, an LDAP search filter like :
>
> `(voPersonSoRID=<CILogon OIDC sub claim value>)`
>
> or
>
> `(voPersonExternalID;app-cilogon=<CILogon OIDC sub claim value>)`
>
> That last example uses LDAP attribute options. See
>
> ```
> https://spaces.at.internet2.edu/display/COmanage/LDAP+Provisioning+Plugin#LDAPProvisioningPlugin-LDAPAttributeOptions
> ```
>
> and
>
> `https://github.com/voperson/voperson/blob/main/voPerson.md#voperson-attribute-options`
>
> Brian Lin and [Scott Koranda] have talked about the OSG CO moving eventually to using LDAP attribute options.
> It will give you more flexibility, 

> > but you do have to be prepared for a few (relatively minor) changes in the details of the LDAP records.

Reviewer: I'm not a fan of the straight up copy-paste here as a final part of the documentation: this should be summarized and distilled to the most relevant points of how the machinery works today, e.g. we don't need references to what things may look like in the future.
